src/Streamly/Internal/Data/Producer.hs

```diff
@@ -0,0 +1,235 @@
+-- stopped in the middle, its state can be extracted, examined, changed, and
+-- then it can be resumed later from the stopped state.
+--
+-- A producer can be used in places where a CPS stream would otherwise be
+-- needed, because the state of the loop can be passed around. However, it can
+-- be much more efficient than CPS because it allows stream fusion and
+-- unecessary function calls can be avoided.
+
+module Streamly.Internal.Data.Producer
+    ( Unfold (..)
+
+    -- * Converting
+    , simplify
+
+    -- * Unfolds
+    , nil
+    , nilM
+    , unfoldrM
+    , fromStreamD
+    , fromList
+
+    -- * Combinators
+    , NestedLoop (..)
+    , concat
+    , parse
+    , parseMany
+    , parseManyD
+    )
+where
+
+#include "inline.hs"
+
+import Control.Exception (assert)
+import Control.Monad.Catch (MonadThrow, throwM)
+import GHC.Exts (SpecConstrAnnotation(..))
+import GHC.Types (SPEC(..))
+import Streamly.Internal.Data.Parser.ParserD (ParseError(..), Step(..))
+import Streamly.Internal.Data.Producer.Source (Source)
+import Streamly.Internal.Data.Stream.StreamD.Step (Step(..))
+import Streamly.Internal.Data.Stream.StreamD.Type (Stream(..))
+import Streamly.Internal.Data.SVar (defState)
+import Streamly.Internal.Data.Unfold.Types (Unfold(..))
+
+import qualified Streamly.Internal.Data.Parser.ParserD as ParserD
+import qualified Streamly.Internal.Data.Parser.ParserK.Types as ParserK
+import qualified Streamly.Internal.Data.Producer.Source as Source
+
+import Streamly.Internal.Data.Producer.Type
+import Prelude hiding (concat)
+
+-- XXX We should write unfolds as producers where possible and define
+-- unfolds using "simplify".
+--
+-------------------------------------------------------------------------------
+-- Converting to unfolds
+-------------------------------------------------------------------------------
+
+-- | Simplify a producer to an unfold.
+--
+-- /Internal/
+{-# INLINE simplify #-}
+simplify :: Producer m a b -> Unfold m a b
+simplify (Producer step inject _) = Unfold step inject
+
+-------------------------------------------------------------------------------
+-- Unfolds
+-------------------------------------------------------------------------------
+
+-- | Convert a StreamD stream into a producer.
+--
+-- /Internal/
+{-# INLINE_NORMAL fromStreamD #-}
+fromStreamD :: Monad m => Producer m (Stream m a) a
+fromStreamD = Producer step return return
+
+    where
+
+    {-# INLINE_LATE step #-}
+    step (UnStream step1 state1) = do
+        r <- step1 defState state1
+        return $ case r of
+            Yield x s -> Yield x (Stream step1 s)
+            Skip s    -> Skip (Stream step1 s)
+            Stop      -> Stop
+
+-------------------------------------------------------------------------------
+-- Parsing
+-------------------------------------------------------------------------------
+
+-- XXX This should probably be moved to the Source module?
```

Comment:
Probably can. It's a low level module anyway. This type is used in a few other places also.

Reply:
`Source` is going through some overhaul in the next commit/PR.

Reply:
Moved.